physical/postgresql.go

```diff
@@ -175,3 +199,143 @@ func (m *PostgreSQLBackend) List(prefix string) ([]string, error) {
 
 	return keys, nil
 }
+
+func (m *PostgreSQLBackend) HAEnabled() bool {
+	return true
+}
+
+// PostgreSQLLock implements the Lock interface for PostgreSQL
+type PostgreSQLLock struct {
+	client *sql.DB
+	table  string
+	leader chan struct{}
+
+	key     string
+	value   string
+	vaultID string
+}
+
+// LockWith initializes a Postgres backend lock
+func (m *PostgreSQLBackend) LockWith(key, value string) (Lock, error) {
+	id, err := uuid.GenerateUUID()
+	if err != nil {
+		return nil, err
+	}
+	return &PostgreSQLLock{
+		client:  m.client,
+		table:   m.lock_table,
+		leader:  make(chan struct{}),
+		key:     key,
+		value:   value,
+		vaultID: id,
+	}, nil
+}
+
+// Lock grabs a lock, or waits until it is available
+func (m *PostgreSQLLock) Lock(stopCh <-chan struct{}) (<-chan struct{}, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		<-stopCh
+		cancel()
+	}()
+	ticker := time.NewTicker(PostgreSQLLockRetryInterval)
+	defer ticker.Stop()
+	for {
+		_, err := m.client.ExecContext(
+			ctx,
+			"DELETE FROM "+m.table+" WHERE expiration < now()",
+		)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
The locking logic here is to use the table's primary key (on column `key`) as a mean to force mutual exclusion. I think server races should not be an issue (the `DELETE` is done only against already expired locks). With this current logic the `DELETE` has to be performed before the `INSERT`, otherwise the `INSERT` may fail in case a previous lock is still present. Could you expand on how exclusion would be done with `INSERT` before `DELETE`?

Changing the transaction isolation level is something I definitely missed. What implementation would you suggest? For all connections from Vault (with the role used to connect to PostgreSQL or at connection time) or for all individual lock-related transactions? I'm very unfamiliar with the performance concerns here.